compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,517 @@
+    CompilerError.invariant(n.path.length > 0, {
+      reason:
+        '[CollectHoistablePropertyLoads] Expected property node, found root node',
+      loc: GeneratedSource,
+    });
+    /**
+     * We add ReactiveScopeDependencies according to instruction ordering,
+     * so all subpaths of a PropertyLoad should already exist
+     * (e.g. a.b is added before a.b.c),
+     */
+    let currNode = this.#getOrCreateRoot(n.identifier);
+    for (let i = 0; i < n.path.length - 1; i++) {
+      currNode = assertNonNull(currNode.properties.get(n.path[i].property));
+    }
+
+    return Tree.#getOrCreateProperty(currNode, n.path.at(-1)!.property);
+  }
+}
+type PropertyLoadInfo = {
+  instr: TInstruction<PropertyLoad>;
+  property: ReactiveScopeDependency;
+};
+
+type TemporariesSidemap = {
+  temporaries: Map<Identifier, Identifier>;
+  properties: Map<Identifier, ReactiveScopeDependency>;
+  propertyLoadInfo: Map<BlockId, Array<PropertyLoadInfo>>;
+};
+
+function collectSidemap(
+  fn: HIRFunction,
+  usedOutsideDeclaringScope: Set<DeclarationId>,
+): TemporariesSidemap {
+  const temporaries = new Map<Identifier, Identifier>();
+  const properties = new Map<Identifier, ReactiveScopeDependency>();
+  const propertyLoadInfo = new Map<BlockId, Array<PropertyLoadInfo>>();
+  for (const [_, block] of fn.body.blocks) {
+    const propertyLoads: Array<PropertyLoadInfo> = [];
+    for (const instr of block.instructions) {
+      const {value, lvalue} = instr;
+      const usedOutside = usedOutsideDeclaringScope.has(
+        lvalue.identifier.declarationId,
+      );
+      if (value.kind === 'PropertyLoad') {
+        const property = getProperty(
+          value.object,
+          value.property,
+          temporaries,
+          properties,
+        );
+        if (!usedOutside) {
+          properties.set(lvalue.identifier, property);
+        }
+        propertyLoads.push({
+          instr: instr as TInstruction<PropertyLoad>,
+          property,
+        });
+      } else if (value.kind === 'LoadLocal') {
+        if (
+          lvalue.identifier.name == null &&
+          value.place.identifier.name !== null &&
+          !usedOutside
+        ) {
+          temporaries.set(lvalue.identifier, value.place.identifier);
+        }
+      }
+    }
+    propertyLoadInfo.set(block.id, propertyLoads);
+  }
+  return {temporaries, properties, propertyLoadInfo};
+}
+
+function collectNodes(
+  fn: HIRFunction,
+  sidemap: TemporariesSidemap,
+): Map<BlockId, BlockInfo> {
+  /**
+   * Due to current limitations of mutable range inference, there are edge cases in
+   * which we infer known-immutable values (e.g. props or hook params) to have a
+   * mutable range and scope.
+   * (see `destructure-array-declaration-to-context-var` fixture)
+   * We track known immutable identifiers to reduce regressions (as PropagateScopeDeps
+   * is being rewritten to HIR).
+   */
+  const knownImmutableIdentifiers = new Set<Identifier>();
```

Comment:
This patches the the regression from fixtures like `destructure-array-declaration-to-context-var` (see [comment on original PR](https://github.com/facebook/react/pull/30079#discussion_r1675390138))